src/libraries/Microsoft.PowerFx.Core/Types/DType.cs

```diff
@@ -1324,10 +1334,15 @@ public DType Add(ref bool fError, DPath path, DName name, DType type)
         public DType Add(DName name, DType type)
         {
             AssertValid();
-            Contracts.Assert(IsAggregate);
+            Contracts.Assert(IsAggregate || IsLazyType);
             Contracts.Assert(name.IsValid);
             type.AssertValid();
 
+            if (IsLazyType)
+            {
+                return new DType(LazyTypeProvider.AddField(name, type), IsTable);
```

Comment:
Field addition and removal are delegated to LazyTypeProvider, so that the table shaping functions don't affect the laziness of the underlying table.